webextension/privileged/ui/CloudDownloadsView.jsm

```diff
@@ -0,0 +1,760 @@
+    return OS.File.exists(path).catch(err => {
+      Cu.reportError(`Couldn't check existance of ${path}`, err);
+      return false;
+    });
+  },
+
+  checkIfExistingCloudProviderDownloadSettings() {
+    // Check if user has previously set preferred download directoy path as
+    // one of cloud provider folder, if yes exit without showing notification
+    // Telemetry - how many such users
+    if (!CloudDownloadsView.providers) {
+      return false;
+    }
+    const dwnldDirPath = this.downloadDirSetting; // await Downloads.getPreferredDownloadsDirectory();
+    const providerValues = [...CloudDownloadsView.providers.values()];
+    return providerValues.some(v => dwnldDirPath.includes(v.downloadPath));
+  },
+
+  async _setDefaultDownloadDirIconURL() {
+    const downloadsDir = await Downloads.getPreferredDownloadsDirectory();
+
+    // Used in defining the correct path to the folder icon.
+    const fph = Services.io.getProtocolHandler("file")
+      .QueryInterface(Ci.nsIFileProtocolHandler);
+
+    this.defaultDownloadDirIconURL = fph.getURLSpecFromFile(new FileUtils.File(downloadsDir));
+  },
+
+  /**
+   * Moves downloaded item to provider download folder and adds
+   * moved file in provider folder as a download item in DownloadList
+   * object shown in user interface in Download Panel and Download History
+   *
+   * @param download
+   *        object representing a single download
+   * @param providerDwnldFldrPath
+   *        String with complete path of provider download folder
+   * @return {Promise} that resolves successfully once download is moved
+   *         to provider folder and movedDownload is added as download item
+   *         in DownloadList object
+   */
+
+  async _moveDownload(download, providerDwnldFldrPath) {
+    // String with complete original target path of download
+    const dwnldTargetPath = download.target.path;
+    let destPath = providerDwnldFldrPath ?
+      OS.Path.join(providerDwnldFldrPath, OS.Path.basename(dwnldTargetPath)) : "";
+
+    // Ensure destPath is a unique file
+    destPath = DownloadPaths.createNiceUniqueFile(new FileUtils.File(destPath)).path;
+
+    try {
+      await OS.File.move(dwnldTargetPath, destPath);
+    } catch (err) {
+      Cu.reportError(err);
+      return;
+    }
+
+    // Create a duplicate download which will show the correct
+    // target path for moved session download in Download UI panel
+    const publicList = await Downloads.getList(Downloads.ALL);
+
+    const movedDownload = await Downloads.createDownload({
+      source: download.source,
+      target: destPath
+    });
+    movedDownload.startTime = download.startTime;
+    movedDownload.succeeded = true;
+    await publicList.add(movedDownload);
+
+    // Update destination path used in download history library panel
+    PlacesUtils.annotations.setPageAnnotation(
+      NetUtil.newURI(download.source.url),
+      "downloads/destinationFileURI",
+      "file://" + destPath, 0,
```

Comment:
This will be incorrect on Windows, and may well have escaping issues, too. You need to determine an actual file: URI here. You can do so by saving the `nsIFile` instance that `createNiceUniqueFile` returns, and passing it to `Services.io.newFileURI`, then fetching its `spec` property.

Reply:
Done